pocld/pocld.cc

```diff
@@ -237,323 +239,575 @@ int listen_rdmacm_events(rdmacm::EventChannelPtr cm_channel,
+               strerror(errno));
+  return nullptr;
+}
+
+void PoclDaemon::readAllClientSocketsThread() {
+  /* keep our listening sockets in the list to streamline the polling code */
+  std::vector<int> open_client_fds = {clients_listen_command_fd,
+                                      clients_listen_stream_fd};
+  std::vector<VirtualContextBase *> socket_contexts = {nullptr, nullptr};
+  std::vector<Request *> incomplete_requests = {nullptr, nullptr};
+  bool fds_changed = true;
+  std::vector<struct pollfd> pfds;
+
+  int pending_client_command = 0;
+  int pending_client_stream = 0;
+
+  while (!exit_helper.exit_requested()) {
+    /* Changes to the list of sockets should be relatively rare so let's
+     * just rewrite the whole thing when it happens; it's a trivial
+     * operation anyway. */
+    if (fds_changed) {
+      pfds.clear();
+      pfds.reserve(open_client_fds.size());
+      for (const int &fd : open_client_fds) {
+        /* Unlike the other error flags POLLRDHUP is only returned if explicitly
+         * polled for */
+        pfds.push_back({fd, POLLIN | POLLRDHUP, 0});
       }
+      fds_changed = false;
     }
-    addr = describe_sockaddr((struct sockaddr *)&client_addr_command, len);
 
-  RESTART2:
-    fd_stream = accept(server_sock_stream,
-                       (struct sockaddr *)&client_addr_stream, &len);
-    if (fd_stream == -1) {
+    /* Just block forever. If/when a socket is closed - including the client
+     * listeners - it triggers a POLLERR/POLLHUP/POLLRDHUP/POLLNVAL. */
+    int num_event_fds = poll(pfds.data(), pfds.size(), -1);
+    POCL_MSG_PRINT_GENERAL("Client socket poll returned %d fds with events\n",
+                           num_event_fds);
+
+    if (num_event_fds < 0) {
       int e = errno;
+      exit_helper.requestExit(strerror(e), e);
+      continue;
+    } else if (num_event_fds == 0) {
+      continue;
+    }
 
-      if (e == EAGAIN)
-        goto RESTART2;
-      else {
-        POCL_MSG_ERR("accept(fd_stream): %s\n", strerror(e));
-        goto EXIT;
+    auto accept_new_connection = [&](struct pollfd &pfd, int bufsize,
+                                     int is_fast, int &hack) {
+      int ev = pfd.revents;
+      pfd.revents = 0; // reset revents to 0 for the next polling round
+      if (ev) {
+        --num_event_fds;
+        if (ev & POLLFD_ERROR_BITS) {
+          exit_helper.requestExit("Client listener socket closed", 0);
+          return false;
+        } else if (ev & POLLIN) {
+          struct sockaddr client_address;
+          socklen_t client_address_length = sizeof(client_address);
+          /* NOTE: address length MUST be initialized to the size of the storage
+           * given as the addr argument */
+          int newfd = accept(pfd.fd, &client_address, &client_address_length);
+          if (newfd > 0) {
+            // XXX: rework (eliminate?) handshakes and just store new fds
+            hack = newfd;
```

Comment:
It's returning the new fd from the function via out parameter so it can be paired with the other fd ("command" and "stream") of the same client. Once requests contain session IDs this is not necessary (and we won't have race conditions when multiple clients try to connect simultaneously)